tests/testsuite/cargo_add/overwrite_public_with_no_public/mod.rs

```diff
@@ -0,0 +1,39 @@
+use cargo_test_support::compare::assert_ui;
+use cargo_test_support::prelude::*;
+use cargo_test_support::Project;
+
+use cargo_test_support::curr_dir;
+
+#[cargo_test]
+fn case() {
+    cargo_test_support::registry::init();
+    for name in ["my-package1", "my-package2"] {
+        for ver in [
+            "0.1.1+my-package",
+            "0.2.0+my-package",
+            "0.2.3+my-package",
+            "0.4.1+my-package",
+            "20.0.0+my-package",
+            "99999.0.0+my-package",
+            "99999.0.0-alpha.1+my-package",
+        ] {
+            cargo_test_support::registry::Package::new(name, ver).publish();
+        }
+    }
+
+    let project = Project::from_template(curr_dir!().join("in"));
+    let project_root = project.root();
+    let cwd = &project_root;
+
+    snapbox::cmd::Command::cargo_ui()
+        .arg("add")
+        .arg_line("my-package1 my-package2@0.4.1 --no-public")
```

Comment:
I don't think these tests are getting anything by adding two packages.  How about we drop it to one?

Yeah, another area where we could clean up the optional tests.

Reply:
same solution : https://github.com/rust-lang/cargo/pull/13046#discussion_r1405740955

Reply:
> By the way, I can make another PR to remove the unnecessary packages and versions for the optional tests. just like https://github.com/rust-lang/cargo/commit/07ba57dd8ba2a4c4f0e4aac8198a3fc69b48e07b

Yes, this would be a great follow up to the previous clean up work in #12189.

In terms of multiple packages ,some tests will want it to verify behavior but we don't need it on every test.  Similar for the many versions.